### DISCUSSION TITLE
#1 Study Introduction (240328)

### DISCUSSION THREAD

**Discussion by MarsJacobs** [efficient-ai-study/efficient-ai-study#16] [2024 Seminar] [answered]
Study Introduction by 김율화/성균관대

**Comment by MarsJacobs** [comment]
이곳에 설문조사의 질문 및 답변이 업데이트 될 예정입니다~

**Comment by YulhwaKim** [comment, marked as answer]
**Q1. AI 하드웨어와 관련된 지식을 쌓고 싶은데, 어떤 방식의 공부를 추천하시나요? 현재 AI 엔지니어로서 소프트웨어 부분에 대한 지식은 어느정도 습득해있고, 어떤 식으로 공부하면 좋을 지 감이 오는데, 하드웨어 쪽은 학교 전공부터도 무관했어서 아예 감이 오지 않아서요. 혹시 강의 영상, 책, 논문, 블로그 등 추천해주실만한 부분이 있다면 공유해주시면 감사하겠습니다.** 
A1. AI 가속기의 역사는 Eyeriss에서부터 시작했다고 해도 과언이 아닌데요, 이런 Eyeriss를 만든 팀에서 만든 AI 하드웨어 관련 [튜토리얼 자료](https://eyeriss.mit.edu/tutorial.html)를 한 번 살펴보시면 좋을 것 같습니다. 그 중에서도 ["Efficient Processing of Deep Neural Networks: A Tutorial and Survey"](https://eems.mit.edu/wp-content/uploads/2017/11/2017_pieee_dnn.pdf)는 DNN 학습 및 추론부터 DNN 가속기 디자인까지 관련 내용을 잘 정리해놨기 때문에 추천드립니다. 2017년에 만들어진 자료라 최신 모델 보다는 CNN 위주로 이야기 하지만, AI 가속기의 핵심이 되는 내용을 자세하게 다루고 있어 AI 하드웨어 입문용으로 좋은 자료라고 생각합니다.
하지만 질문자께서 소프트웨어 쪽에 훨씬 일가견이 있으신 분인 것 같아, CUDA programming 및 GPU 아키텍처를 공부하는 쪽을 좀 더 추천 드립니다. 잘 아시겠지만, GPU는 AI모델 학습/추론에 있어서 빠질 수 없는 핵심 AI하드웨어이며, CUDA는 NIVIDA GPU를 위한 programming 언어 입니다. CUDA 프로그래밍은 GPU 아키텍처에 대한 이해 없이는 할 수 없으므로, GPU 아키텍처도 공부하면서 CUDA 프로그래밍을 병행해보시면 어떨까 해요. CUDA 프로그래밍 입문을 위해서는 2010년에 나온 "예제로 배우는 CUDA 프로그래밍(CUDA by Example), Jason Sanders & Edward Kandrot"를 추천 드립니다. 책이 부담스럽다면 NVIDIA에서 발표한 [CUDA programming introduction 슬라이드](https://www.olcf.ornl.gov/wp-content/uploads/2013/02/Intro_to_CUDA_C-TS.pdf)를 참고하는 것도 추천드립니다. GPU 아키텍처와 관련된 핵심 용어 정리는 [링크를 걸어둔 블로그](https://www.cherryservers.com/blog/everything-you-need-to-know-about-gpu-architecture#what-is-a-gpu)에 잘 정리되어 있는 것 같고, 전반적인 GPU 구조 및 구동 방식은 [CMU의 Kayvon Fatahalian교수님 수업자료](https://www.cs.cmu.edu/afs/cs/academic/class/15462-f11/www/lec_slides/lec19.pdf)에 잘 정리되어 있는 것 같습니다. 기본 튜토리얼 자료로 입문을 한 후, 최신 CUDA 프로그래밍 룰 및 GPU 아키텍처는 [CUDA C++ Programming Guide](https://docs.nvidia.com/cuda/cuda-c-programming-guide/)와 최신 아키텍처(Volta, Ampere, Hopper,...)의 White Paper를 읽어보시는 걸 추천드립니다. 
CUDA 프로그래밍은 소프트웨어와 하드웨어 지식 모두 필요로 하는 부분이라 CUDA 엔지니어는 수요에 비해 수가 많이 부족한 걸로 알고 있습니다. 이쪽을 공부해보시면 AI하드웨어에 대한 이해도도 높이고, 질문해주신 분의 소프트웨어적인 강점도 살릴 수 있지 않을까 싶습니다.

**Q2. W4A4 연산을 할때 보통 A32에 accumulationa을 한다고들 많이 이야기 하던데 혹시 이 연산이 실제로 어떻게 실행되는지 알 수 있을까요? W4A4를 곱하는 순간 overflow가 날 것 같은데 어떻게 문제 없이 계산이 진행되는지 궁금합니다. 해당 정리된 곳을 알려주셔도 충분할 것 같습니다.**
A2. 4bit multiplication을 수행한다는 건 입력 데이터의 비트 수가 4bit라는 것을 의미하며, 연산 결과는 해당 연산에서 나올 수 있는 모든 경우를 다룰 수 있도록 비트 수가 조절되게 됩니다. 예를 들면, [4bit multiplier](https://www.researchgate.net/profile/Navdeep-Goel/publication/266890236/figure/fig4/AS:670347519074307@1536834711199/4-bit-Array-Multiplier-AM.ppm)는 8bit 연산 결과를 도출해내도록 디자인됩니다. 따라서 질문주신 바와 같이 W4A4, 즉 4-bit weight와 activation간의 곱셈 결과는 하드웨어 상에서 8-bit로 overflow 없이 처리됩니다. multiplication 이후 accumulation 단계에서는 질문자분께서 이야기 해주신 것처럼 overflow를 막기 위해 넉넉하게 32-bit accumulator를 이용해 여러 곱셈 결과를 합치게 됩니다. (합쳐야 하는 8-bit 데이터 개수가 2^(32-8)개 이하인 경우 32-bit accumulator에서 overflow가 나지 않습니다.)

**Q3. 혹시, 저처럼 초보자에게 도움이 될만한 경량화 관련 도서가 있을까요?**
A3. Survey 논문을 한 번 읽어보시는 게 어떨까 합니다. 개인적으로 ["Pruning and Quantization for Deep Neural Network Acceleration: A
Survey"](https://arxiv.org/pdf/2101.09671.pdf)를 추천드립니다. 다양한 pruning 및 quantization 기법을 카테고리 별로 잘 정리해 두어 입문용으로 좋은 것 같습니다. 그리고 ["awesome-model-quantization"](https://github.com/htqin/awesome-model-quantization)이라는 github 페이지가 quantization과 관련된 논문 및 자료를 잘 정리해두고 있습니다. 논문 수가 다소 많아 다 읽어보기는 어려울 듯 하고, 여기서 추천하는 survey 논문 정도는 한 번씩 읽어보면 좋지 않을까 해요.
